xls/examples/gcd.x

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import std
+
+// https://en.wikipedia.org/wiki/Greatest_common_divisor#Euclidean_algorithm
+fn gcd_euclidean<N: u32, DN: u32 = {N * u32:2}>(a: uN[N], b: uN[N]) -> uN[N] {
+    let (_, gcd) = for (i, (a, b)) in range(u32:0, DN) {
+        let (d, r) = std::iterative_binary_div(a, b);
```

Comment:
should move this in the else case, so that we don't perform unnecessary division when the remainder is `0`.

Reply:
Not sure I understand. How can this move into the else case if the result of the computation is the condition?

Reply:
I think gcd is a good candidate for proc implementations, which would let you do the "early return" I think you're describing. Each proc tick would do one division (or `gcd_binary_match`), and you send the output when you reach the stop condition.

Fine to leave as a TODO, and these function implementations are nice because they work well w/ quickcheck.

Reply:
> Not sure I understand. How can this move into the else case if the result of the computation is the condition?

@meheff since we always pass `r` to the next iteration, I was thinking we could `sel` based on the result of the previous division, rather than always performing the same division again and again until the end of the loop.

Reply:
> I think gcd is a good candidate for proc implementations, which would let you do the "early return" I think you're describing.

It could be interesting to compare a proc implementation to the automatically pipelined version.

Reply:
> @meheff since we always pass `r` to the next iteration, I was thinking we could `sel` based on the result of the previous division, rather than always performing the same division again and again until the end of the loop.

As discussed simplified the implementation to only trigger div_mod in `for` recurrence w/ a non-zero modulo.

This apparently result in a reduced critical path and bom, see https://colab.research.google.com/gist/proppy/01142e33a2be8b71d6b317ee4ce7fde4/xls-playground-sandbox.ipynb